source/Patches/AmongUsClient_OnGameEnd.cs

```diff
@@ -10,8 +10,29 @@ namespace TownOfUs
     [HarmonyPatch(typeof(EndGameManager), nameof(EndGameManager.Start))]
     public class EndGameManager_SetEverythingUp
     {
+        // Implement a method to record the roles assigned to each player at the end of a game.
+        // Store the role history in a session-based collection.
+        private static Dictionary<byte, List<RoleEnum>> _sessionRoleHistory = new Dictionary<byte, List<RoleEnum>>();
+
+        public static void RecordSessionRoles() {
+            foreach (var player in PlayerControl.AllPlayerControls) {
+                var playerId = player.PlayerId;
+                var roleType = Role.GetRole(player)?.RoleType ?? RoleEnum.None;
+
+                if (!_sessionRoleHistory.ContainsKey(playerId)) {
+                    _sessionRoleHistory[playerId] = new List<RoleEnum>();
+                }
+
+                _sessionRoleHistory[playerId].Add(roleType);
+            }
+        }
+
         public static void Prefix()
         {
+          if (CustomGameOptions.WeightedRoleSelection)  
+          { 
+              RecordSessionRoles(); // Record roles at the end of each game
+          }
```

Comment:
This should probably happen at the start of the game and not the end as the final roles people end up with are not the same as those they initially were assigned and we only care about those original roles.

Additionally, I'm not sure how PlayerIds are handled in Among Us so doing it at game start may handle situations like disconnects better?